Add utility functions to rename a file or folder src to dst with retrying.

loonghao opened the pull request:
This PR is try to fix #1665

JeanChristopheMorinPerso commented:
Thanks @loonghao for submitting this PR. My main concern right now is speed. Starting a subprocess on Windows usually adds a noticeable overhead.

Could the code first try to `os.rename` and only fallback to the subprocess method if it's running on windows and the error is "WinError 5" or something like that?

Also, it would be great if you could add tests to cover all cases covered by the change.

loonghao commented:
@JeanChristopheMorinPerso I have updated the code and added a unit test for this function. Please take another look. Thanks.

JeanChristopheMorinPerso commented on src/rez/utils/filesystem.py:
```diff
@@ -690,3 +692,56 @@ def windows_long_path(dos_path):
+    try:
+        os.rename(src, dst)
+    except PermissionError:
+        if is_windows and which("robocopy") and os.path.isdir(src):
```
The exception will be eaten by the except block if this condition is not true.

JeanChristopheMorinPerso commented on src/rez/tests/test_utils.py:
```diff
@@ -48,3 +50,20 @@ def test_unix_case_insensistive_platform(self):
+    def test_rename_folder(self):
+        src_path = tempfile.mkdtemp("src_path")
+        dst_path = "{0}_new".format(src_path)
+        if platform_.name == 'windows':
```
I would create two different test functions instead of an if. Also we should test the non failing scenario and we also need to test the scenario where everything fails.

JeanChristopheMorinPerso reviewed (changes requested):
Thanks! There is a bug in the code and I think we can add more tests.

loonghao commented:
> Thanks! There is a bug in the code and I think we can add more tests.

@JeanChristopheMorinPerso I have updated the code and added more unit tests. Please take another look. 

Thanks.

JeanChristopheMorinPerso commented on src/rez/utils/filesystem.py:
```diff
@@ -690,3 +692,58 @@ def windows_long_path(dos_path):
+
+
+def rename(src, dst):
+    """Utility functions to rename a file or folder src to dst with retrying.
```
```suggestion
    """Utility function to rename a file or folder src to dst with retrying.
```

JeanChristopheMorinPerso commented on src/rez/utils/filesystem.py:
```diff
@@ -690,3 +692,58 @@ def windows_long_path(dos_path):
+def rename(src, dst):
+    """Utility functions to rename a file or folder src to dst with retrying.
+
+    This function uses the built-in 'os.rename()' method, but in case of a
+    'PermissionError', it uses the 'robocopy' command line utility on Windows.
+    This is done to avoid the frequent 'Access is denied' exception raised by 'os.rename()' on Windows.
```
```suggestion
    This function uses the built-in 'os.rename()' function and falls back to 'robocopy' tool if os.rename raises a PermissionError exception.
```

JeanChristopheMorinPerso commented on src/rez/utils/filesystem.py:
```diff
@@ -690,3 +692,58 @@ def windows_long_path(dos_path):
+    Raises:
+        OSError: If renaming fails after all attempts.
+
+    References:
+        - https://github.com/conan-io/conan/blob/develop2/conan/tools/files/files.py#L207
+
+    """
```
```suggestion
    """
    # Inspired by https://github.com/conan-io/conan/blob/2.1.0/conan/tools/files/files.py#L207
```

JeanChristopheMorinPerso commented:
Thanks @loonghao ! This should hopefully be my last batch of comments.

loonghao commented:
> Thanks @loonghao ! This should hopefully be my last batch of comments.

@JeanChristopheMorinPerso I have made all the changes based on your feedback and addressed all the comments. Could you please take another look. Thanks!

JeanChristopheMorinPerso reviewed (approved):
Looks good to me. Thanks for taking the time to fix this and address all my comments! If everything goes as planned, this will be released in the next release.
